### ISSUE TITLE
How about @DynamicProxyHint and @SubstitutionHint

### ISSUE THREAD

**Issue by linux-china** [GoodforGod/graalvm-hint#2]
@GoodforGod, it's nice project  👍   

Now I use optionNames to add DynamicProxy support with following code:  

```
@NativeImageHint(optionNames = {"-H:DynamicProxyConfigurationResources=com.example.UserService"})
```

Another problem is   `-H:SubstitutionResources=${.}/substitutions.json` , and substitutions should be supplied by json file.  

How about new @DynamicProxyHint and  @SubstitutionHint annotations?  I think they are more friendly to developers.

**Comment by GoodforGod** [comment]
Couldn't find any info about SubstitutionResources extent [this](https://www.graalvm.org/22.0/reference-manual/native-image/BuildConfiguration/)

This option basically only allow to specify resources like list of string to be substituted?
Like:
`  -H:SubstitutionResources=${.}/substitutions1.json,${.}/substitutions1.json`

But what exactly they substitute? What is their purpose? I dont get it and couldn't find any documentation, may be you can provide some info please
@linux-china

**Comment by linux-china** [comment]
Yes, it's hard to find document, and you can find some examples from following:

*  https://github.com/spring-projects-experimental/spring-native/tree/main/spring-native/src/main/java/org/springframework/nativex/substitutions  
* https://github.com/panga/hammock-graal  
* https://blog.frankel.ch/solving-substitution-graalvm-issue/

**Comment by GoodforGod** [comment]
Thanks! Looks like some wodo black magic, will try to find some more info from GraalVM devs

**Comment by GoodforGod** [comment]
Can you please try DynamycProxyHint, I deploy `0.18.0-SNASHOT` version and give your feedback and potential bugs, cause I don't have any real world usages for this hint
@linux-china 

Do not forget to add snapshot repo
```
repositories {
    maven { url "https://oss.sonatype.org/content/repositories/snapshots" }
}
```

**Comment by GoodforGod** [comment]
Looks like Substitution option is deprecated so, no point in supporting it

https://github.com/oracle/graal/commit/6054669045f1504ccfbdf987f2200b365335de96

**Comment by linux-china** [comment]
@GoodforGod  yes,   from https://www.graalvm.org/release-notes/22_0/.   Removed the option -H:SubstitutionFiles=... to register substitutions via a JSON file.

But you can find some information from https://docs.oracle.com/en/graalvm/enterprise/22/docs/reference-manual/native-image/BuildConfiguration/

![image](https://user-images.githubusercontent.com/46711/162487019-a870ce49-90ce-4cb2-9c75-c67088a8313a.png)

**Comment by GoodforGod** [comment]
Looks like just devs forgot to clean up this doc, cause I asked them and they confirmed option is deprecated

**Comment by linux-china** [comment]
@GoodforGod  only few developers use substitution feature, and it's not for normal developers.     @DynamycProxyHint  is good enough for me now :)

**Comment by GoodforGod** [comment]
Would like a lot to hear your feedback on @DynamicProxyHint, is it working properly, to you like the @DynamicProxyHint.Configuration hint annotation, so please share your experience when you will play with it :)

**Comment by GoodforGod** [comment]
Hello @linux-china any success trying `@DynamicProxyHint` annotation?:)

**Comment by linux-china** [comment]
@GoodforGod  It works, and I forgot to give you feedback.    

```
@DynamicProxyHint
interface UserService {

    String findNickById(Integer id);
}

class UserServiceHandler implements InvocationHandler {

    @Override
    public Object invoke(Object proxy, Method method, Object[] args) throws Throwable {
        return "nick: 1";
    }
}

```

**Comment by GoodforGod** [comment]
Great! Thanks for feedback!

Actually it is merged, but there is slight change, the artifact name is always `hint` now and package name is same as class package name or default one, so the path for hints is something like:
`"META-INF/native-image/" + package + "/hint/" + fileName;`

**Comment by linux-china** [comment]
@GoodforGod  I got the following error if without package declare. 

```
error: Couldn't write META-INF/native-image//hint/dynamic-proxy-config.json due to: Invalid relative name: META-INF/native-image//hint/dynamic-proxy-config.json
error: Couldn't write META-INF/native-image//hint/native-image.properties due to: Invalid relative name: META-INF/native-image//hint/native-image.properties
error: Couldn't write META-INF/native-image//hint/reflect-config.json due to: Invalid relative name: META-INF/native-image//hint/reflect-config.json
3 errors
```

**Comment by GoodforGod** [comment]
Just published `0.18.0-SNAPSHOT` with hotfix for this issue, please refresh dependencies in gradle/maven and try again, hope issue will be resolved :)

**Comment by linux-china** [comment]
Yes, it works well  without package declare now.
